server/src/main/java/org/apache/kafka/server/share/metrics/SharePartitionMetrics.java

```diff
@@ -0,0 +1,166 @@
+
+import org.apache.kafka.common.utils.Utils;
+import org.apache.kafka.server.metrics.KafkaMetricsGroup;
+
+import com.yammer.metrics.core.Histogram;
+import com.yammer.metrics.core.Meter;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+/**
+ * SharePartitionMetrics is used to track the broker-side metrics for the SharePartition.
+ */
+public class SharePartitionMetrics implements AutoCloseable {
+
+    public static final String INFLIGHT_MESSAGE_COUNT = "InFlightMessageCount";
+    public static final String INFLIGHT_BATCH_COUNT = "InFlightBatchCount";
+
+    private static final String ACQUISITION_LOCK_TIMEOUT_PER_SEC = "AcquisitionLockTimeoutPerSec";
+    private static final String INFLIGHT_BATCH_MESSAGE_COUNT = "InFlightBatchMessageCount";
+    private static final String FETCH_LOCK_TIME_MS = "FetchLockTimeMs";
+    private static final String FETCH_LOCK_RATIO = "FetchLockRatio";
+
+    /**
+     * Metric for the rate of acquisition lock timeouts for records.
+     */
+    private final Meter acquisitionLockTimeoutPerSec;
+    /**
+     * Metric for the number of in-flight messages for the batch.
+     */
+    private final Histogram inFlightBatchMessageCount;
+    /**
+     * Metric for the time the fetch lock is held.
+     */
+    private final Histogram fetchLockTimeMs;
+    /**
+     * Metric for the ratio of fetch lock time to the total time.
+     */
+    private final Histogram fetchLockRatio;
+
+    private final Map<String, String> tags;
+    private final KafkaMetricsGroup metricsGroup;
+
+    public SharePartitionMetrics(String groupId, String topic, int partition) {
+        this.tags = Utils.mkMap(
+            Utils.mkEntry("group", Objects.requireNonNull(groupId)),
+            Utils.mkEntry("topic", Objects.requireNonNull(topic)),
+            Utils.mkEntry("partition", String.valueOf(partition))
+        );
+        this.metricsGroup = new KafkaMetricsGroup("kafka.server", "SharePartitionMetrics");
+
+        this.acquisitionLockTimeoutPerSec = metricsGroup.newMeter(
+            ACQUISITION_LOCK_TIMEOUT_PER_SEC,
+            "acquisition lock timeout",
+            TimeUnit.SECONDS,
+            this.tags);
+
+        this.inFlightBatchMessageCount = metricsGroup.newHistogram(
+            INFLIGHT_BATCH_MESSAGE_COUNT,
+            true,
+            this.tags);
+
+        this.fetchLockTimeMs = metricsGroup.newHistogram(
+            FETCH_LOCK_TIME_MS,
+            true,
+            this.tags);
+
+        this.fetchLockRatio = metricsGroup.newHistogram(
+            FETCH_LOCK_RATIO,
+            true,
+            this.tags);
+    }
+
+    /**
+     * Register a gauge for the in-flight message count.
+     *
+     * @param messageCountSupplier The supplier for the in-flight message count.
+     */
+    public void registerInflightMessageCount(Supplier<Long> messageCountSupplier) {
```

Comment:
`registerInFlight...` please. And the next method too.

Reply:
Done.